record-layer.go

```diff
@@ -318,28 +356,47 @@ func (r *RecordLayer) nextRecord() (*TLSPlaintext, error) {
 	pt.fragment = make([]byte, size)
 	copy(pt.fragment, body)
 
+	// TODO(ekr@rtfm.com): Enforce that for epoch > 0, the content type is app data.
+
 	// Attempt to decrypt fragment
-	if cipher.cipher != nil {
-		seq := cipher.seq
-		if r.datagram {
-			seq = header[3:11]
-		}
-		// TODO(ekr@rtfm.com): Handle the wrong epoch.
+	seq := cipher.seq
+	if r.datagram {
 		// TODO(ekr@rtfm.com): Handle duplicates.
-		logf(logTypeIO, "RecordLayer.ReadRecord epoch=[%s] seq=[%x] [%d] ciphertext=[%x]", cipher.epoch.label(), seq, pt.contentType, pt.fragment)
+		seq, _ = decodeUint(header[3:11], 8)
+		epoch := Epoch(seq >> 48)
+
+		// Look up the cipher suite from the epoch
+		if epoch != cipher.epoch {
+			logf(logTypeIO, "%s Message from non-current epoch: [%v != %v]", r.label, epoch,
+				cipher.epoch)
+			if !allowOldEpoch {
```

Comment:
It seems like you don't know at this point whether the epoch is old or just bogus.  Maybe reorder the invalid vs. old check and use the knowledge of the epoch as a sign that it's old, not bogus.